### DISCUSSION TITLE
sudo: better control over timestamps

### DISCUSSION THREAD

**Discussion by Kentzo** [Homebrew/discussions#4207] [Tap maintenance and Homebrew development]
Currently `brew install` quietly leaks sudo authentication from formula to formula. This effect may be undesired and is the most vivid when installing a combination of multiple casks and formulas previously never tried by the user.

I'd like to propose the following behavior:

1. If `brew install` is being run from a shell with pre-authenticated sudo as reported by `sudo -nv` (or `sudo -Nv` in newer versions) then maintain that authentication throughout the execution no matter how long the command takes or the number of formulas requested
2. Otherwise clear cached credentials via `sudo --remove-timestamp` after each installed formula requiring the user to re-authenticate

This behavior should be optional, for the time being, controlled by the `HOMEBREW_SUDO_REMOVETIMESTAMP`.

**Comment by SMillerDev** [comment]
What would the benefit of this behaviour be?

**Reply by Kentzo** [reply]
Explicit control over whether I want a cask, that I'm installing for the first time, to run an arbitrary script with root privileges. When I encounter such casks, my personal preference is to interrupt installation and audit the downloaded package.

**Reply by p-linnane** [reply]
This does not seem like something the vast majority of Homebrew users would use. Given that we as maintainers have limited time to improve Homebrew, this does not seem likely to happen.

**Reply by Kentzo** [reply]
Do you have any objection against the feature as a whole? If I provide the initial implementation, then will you may consider for inclusion?

**Reply by p-linnane** [reply]
Feel free to open a PR in https://github.com/Homebrew/brew with an initial implementation to receive feedback.
